### PR TITLE
Ability to decouple alps::params from alps::hdf5

### PR DESCRIPTION
**Pull request by galexv**
This PR conditionally decouples `alps::params` from `alps::hdf5`. Specifically, if ALPSCore is built with
`-DALPS_MODULES_DISABLE=hdf5`, the `params` module will be built without archive support. This is useful if all I need from ALPSCore is its `utilities` and `params` modules.

The disadvantage is that the version of ALPSCore build like this is not compatible with the "regular" ALPSCore (e.g., it has different definition of the `alps::params` class).

This PR partially addresses issue #572.

### CODE REVIEW

**Comment by egull** [comment]
Markus, this is near and dear to you. Does it help you?

**Comment by mwallerb** [comment]
@galexv Quick question: is with/without HDF really ABI-incompatible?  Usually, adding methods to a class should not break that, or am I wrong?  If so, would making the methods `virtual` help?

**Comment by galexv** [comment]
@mwallerb (I should have said more clearly that it's a semi-measure: i do need ALPSCore's params in a different project and i do not want to carry anything else with it).

I am not sure about binary incompatibility, but certainly [ODR](https://en.cppreference.com/w/cpp/language/definition#One_Definition_Rule) is violated: an application compiled with the regular ALPSCore should not use "lightweighted" ALPSCore libraries, because the app and the lightweight libraries used different definition of the `params` class in their headers.

Could you clarify possible concerns --- in what sense you need the binary compatibility? 

For example, if a user's app needs the lightweight ALPSCore, the user can recompile ALPSCore excluding the unwanted modules and link to the application. The drawback is that there could be a proliferation of local ALPSCore versions each with its own set of compiled modules, which potentially may lead to confusion. That's why this PR is only a partial solution.

**Comment by mwallerb** [comment]
@galexv OK, maybe I typed too fast.  Anyway, perhaps as a tangent, a way to *actually* decouple parameters from HDF5 would be to negotiate it via a purely abstract base class. 

Take a look at the ALEA's [serializer] interface and it's [HDF5 implementation]: it exposes exactly as much of HDF5 as needed for the serialization of arrays to disk.  Crucially though, the HDF5 dependency is broken in compilation and linking, because the connection is only made in the user code.

I think this has a number of advantages to the preprocessor switches: it is less brittle in the case of a mixed setup and it also improves compilation time.  In theory you have to pay a price for dispatch, but since most HDF5 calls involve I/O this should not matter.

I don't know the parameter part of ALPS well, so I have difficulty estimating the effort ... but at least there is some prior work in ALEA that one can rely on.  So here's the proposal:

 1. elevate `alea::serializer` and `alea::deserializer` to common ALPS classes.
 2. specialize `serializer::write()` and `deserializer::read()` for types needed by parameters
 3. add factory methods to `hdf5_archive` class that return `serializer` / `deserializer` instances.  Something to the effect of:

         class hdf5_archive {
              class hdf5_serializer : public serializer { ... };
              class hdf5_deserializer : public deserializer { ... };
              ...
              hdf5_serializer get_serializer();
              hdf5_deserializer get_deserializer() const; // not sure about const
         }

 4. Template `save()` and `load()` methods of parameters over archive type, call the factory method to obtain serializer, pass a reference the underlying ABC to the compiled code.
 5. Profit!

What do you think?

To summarize, I'm totally okay with the current approach if manpower is at a premium, but above would be The Proper Way<sup>TM</sup> to do it IMHO.

[serializer]: https://github.com/ALPSCore/ALPSCore/blob/master/alea/include/alps/alea/core.hpp#L320
[HDF5 implementation]: https://github.com/ALPSCore/ALPSCore/blob/master/alea/include/alps/alea/hdf5.hpp#L18

**Comment by egull** [comment]
I agree with that. There are a bunch of places where interfaces rather than dependencies will help a lot – this is one of them. Alex, we can discuss this on Monday...

**Comment by galexv** [comment]
@mwallerb @egull I agree with this assessment. We can discuss it further on Monday. I'd still think that this PR may be merged as a temporary solution.
